File: post-service/src/main/java/com/blogit/post/entity/Post.java
package com.blogit.post.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "posts")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Post {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long userId;

    @Column(nullable = false, length = 500)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PostVisibility visibility = PostVisibility.PUBLIC;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "post_hashtags", joinColumns = @JoinColumn(name = "post_id"))
    @Column(name = "hashtag")
    @Builder.Default
    private Set<String> hashtags = new HashSet<>();

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "post_media", joinColumns = @JoinColumn(name = "post_id"))
    @Column(name = "media_url")
    @Builder.Default
    private Set<String> mediaUrls = new HashSet<>();

    @Column(nullable = false)
    @Builder.Default
    private Long likesCount = 0L;

    @Column(nullable = false)
    @Builder.Default
    private Long commentsCount = 0L;

    @Column(nullable = false)
    @Builder.Default
    private Long sharesCount = 0L;

    @Column(nullable = false)
    @Builder.Default
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @Version
    private Long version;

    public enum PostVisibility {
        PUBLIC, FOLLOWERS_ONLY, PRIVATE
    }

    // Helper methods
    public void incrementLikesCount() {
        this.likesCount++;
    }

    public void decrementLikesCount() {
        if (this.likesCount > 0) {
            this.likesCount--;
        }
    }

    public void incrementCommentsCount() {
        this.commentsCount++;
    }

    public void decrementCommentsCount() {
        if (this.commentsCount > 0) {
            this.commentsCount--;
        }
    }

    public void incrementSharesCount() {
        this.sharesCount++;
    }

    public void addHashtag(String hashtag) {
        this.hashtags.add(hashtag.toLowerCase());
    }

    public void addMediaUrl(String mediaUrl) {
        this.mediaUrls.add(mediaUrl);
    }
}

File: blogit-frontend/src/types/post.ts
export interface CreatePostRequest {
  title: string;
  content: string;
  visibility: PostVisibility;
  hashtags?: string[];
  mediaUrls?: string[];
}

export interface UpdatePostRequest {
  title?: string;
  content?: string;
  visibility?: PostVisibility;
  hashtags?: string[];
  mediaUrls?: string[];
}

export interface PostResponse {
  id: string;
  userId: string;
  title: string;
  content: string;
  visibility: PostVisibility;
  hashtags: string[];
  mediaUrls: string[];
  likesCount: number;
  commentsCount: number;
  sharesCount: number;
  createdAt: Date;
  updatedAt: Date;
  author?: {
    id: string;
    username: string;
    fullName: string;
    profileImage?: string;
  };
  isLiked?: boolean;
}

export enum PostVisibility {
  PUBLIC = 'PUBLIC',
  FOLLOWERS_ONLY = 'FOLLOWERS_ONLY',
  PRIVATE = 'PRIVATE'
}

export interface PostStats {
  likesCount: number;
  commentsCount: number;
  sharesCount: number;
  viewsCount: number;
  engagementRate: number;
}

export interface DraftPost extends CreatePostRequest {
  id: string;
  createdAt: Date;
  updatedAt: Date;
} 